pkg/chartsync/git.go

```diff
@@ -0,0 +1,407 @@
+import (
+	"context"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/weaveworks/flux/git"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/fluxcd/helm-operator/pkg/apis/helm.fluxcd.io/v1"
+	clientset "github.com/fluxcd/helm-operator/pkg/client/clientset/versioned"
+	lister "github.com/fluxcd/helm-operator/pkg/client/listers/helm.fluxcd.io/v1"
+	"github.com/fluxcd/helm-operator/pkg/status"
+)
+
+// Condition change reasons
+const (
+	ReasonGitNotReady = "GitNotReady"
+	ReasonGitCloned   = "GitRepoCloned"
+)
+
+// ReleaseQueue is an add-only `workqueue.RateLimitingInterface`
+type ReleaseQueue interface {
+	AddRateLimited(item interface{})
+}
+
+// GitConfig holds the configuration for git operations.
+type GitConfig struct {
+	GitTimeout      time.Duration
+	GitPollInterval time.Duration
+}
+
+// GitChartSource holds the source references to a Helm chart in git,
+// and a minimal working clone.
+type GitChartSource struct {
+	sync.Mutex
+	Export *git.Export
+	Mirror string
+	Remote string
+	Ref    string
+	Head   string
+}
+
+// ChartPath returns the absolute chart path for the given relative
+// path.
+func (c *GitChartSource) ChartPath(relativePath string) string {
+	return filepath.Join(c.Export.Dir(), relativePath)
+}
+
+// forHelmRelease returns true if the given `v1.HelmRelease`s
+// `v1.GitChartSource` matches the GitChartSource.
+func (c *GitChartSource) forHelmRelease(hr *v1.HelmRelease) bool {
+	if hr == nil || hr.Spec.GitChartSource == nil {
+		return false
+	}
+	return c.Mirror == mirrorName(hr) && c.Remote == hr.Spec.GitURL && c.Ref == hr.Spec.Ref
+}
+
+// GitChartSourceSync syncs `GitChartSource`s with their mirrors.
+type GitChartSourceSync struct {
+	logger log.Logger
+	config GitConfig
+
+	lister lister.HelmReleaseLister
+	client clientset.Interface
+
+	mirrors *git.Mirrors
+
+	sourcesMu sync.Mutex
+	sources   map[string]*GitChartSource
+
+	releaseQueue ReleaseQueue
+}
+
+func NewGitChartSourceSync(logger log.Logger,
+	lister lister.HelmReleaseLister, client clientset.Interface, cfg GitConfig, queue ReleaseQueue) *GitChartSourceSync {
+
+	return &GitChartSourceSync{
+		logger:       logger,
+		config:       cfg,
+
+		lister:       lister,
+		client:       client,
+
+		mirrors:      git.NewMirrors(),
+		sourcesMu:    sync.Mutex{},
+		sources:      make(map[string]*GitChartSource),
+
+		releaseQueue: queue,
+	}
+}
+
+// Run starts the sync of `GitChartSource`s.
+func (c *GitChartSourceSync) Run(stopCh <-chan struct{}, errCh chan error, wg *sync.WaitGroup) {
+	c.logger.Log("info", "starting sync of git chart sources")
+
+	wg.Add(1)
+	go func() {
+		defer runtime.HandleCrash()
+		defer func() {
+			c.mirrors.StopAllAndWait()
+			wg.Done()
+		}()
+
+		for {
+			select {
+			case changed := <-c.mirrors.Changes():
+				for mirrorName := range changed {
+					mirror, ok := c.mirrors.Get(mirrorName)
```

Comment:
This function is pretty large, I would extract the body of the for loop into a separate method (e.g. `processChangedMirror()`)

Reply:
Maybe it's better to only factor out the `ok == true` part of the body.